### PR TITLE
#6766: MapStore crashes when importing a 4326 layer with data beyond the area of use of 3857

### PR DESCRIPTION
**Pull request by MuhweziDeo**
## Description
Adding an EPSG:4326 layer containing data beyond the area of use of EPSG:3857 makes something wrong happen in MapStore:

- the data is not drawn on the map
- trying to switch the CRS of the map to 4326 crashes MapStore

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/geosolutions-it/MapStore2/blob/master/CONTRIBUTING.md
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

**What kind of change does this PR introduce?** (check one with "x", remove the others)
 - [x] Bugfix

## Issue

**What is the current behavior?**

- Nothing is initially shown
- A crash happens trying to switch the map's CRS

#6766

**What is the new behavior?**
A basic solution would be to warn the user about the problem with data boundaries beyond the accepted boundaries for the currently selected CRS (probably with less technical wording!) and block the upload.

## Breaking change
**Does this PR introduce a breaking change?** (check one with "x", remove the other)
 - [ ] Yes, and I documented them in migration notes
 - [x] No

### CODE REVIEW

File: web/client/localConfig.json [hunk 1]
```diff
@@ -419,7 +419,7 @@
                   "EPSG:3857"
                 ],
                 "allowedRoles": [
-                  "ADMIN"
+                  "ALL"
```
Comment (MV88) [inline web/client/localConfig.json:422]: why this change? please, revert this

File: web/client/translations/data.en-US.json [hunk 2]
```diff
@@ -1346,7 +1346,8 @@
             "errors": {
                 "fileNotSupported": "File not supported",
                 "unknownError": "there was an unknown error during import",
-                "projectionNotSupported": "The projection of the file(s) you're trying to import is not supported"
+                "projectionNotSupported": "The projection of the file(s) you're trying to import is not supported",
+                "fileBeyondBoundaries": "The file  {filename} couldnot be imported because it doesnot fit inside the map boundaries"
```
Comment (MV88) [inline web/client/translations/data.en-US.json:1350]:
```suggestion
                "fileBeyondBoundaries": "The file  {filename} cannot be imported because it does not fit the map boundaries"
```

File: web/client/components/import/dragZone/enhancers/useFiles.js [hunk 3]
```diff
@@ -1,5 +1,29 @@
...
+    return getProjections().find(project => project.code === code) || {extent: [-20026376.39, -20048966.10, 20026376.39, 20048966.10]};
+};
+
+// checks if a layer fits within the max extent
+export const checkIfLayerFitsExtentForProjection = layer => {
+    const crs = layer.bbox?.crs || "EPSG:3857";
+    const { extent } = getExtentForProjection(crs);
+    const [, , maxX, maxY] = turfBbox({type: 'FeatureCollection', features: layer.features || []});
+    return ((extent[2] >= maxX) && (extent[3] >= maxY));
+};
```
Comment (MV88) [inline web/client/components/import/dragZone/enhancers/useFiles.js:26]: for each utility function please :
- move it to CoordinatesUtils.js
- add a proper jsdoc
- add unit test

File: web/client/components/import/dragZone/enhancers/useFiles.js [hunk 4]
```diff
@@ -1,5 +1,29 @@
...
+// finds a projection from projectionDefs
+const getExtentForProjection = code => {
+    return getProjections().find(project => project.code === code) || {extent: [-20026376.39, -20048966.10, 20026376.39, 20048966.10]};
+};
+
+// checks if a layer fits within the max extent
+export const checkIfLayerFitsExtentForProjection = layer => {
+    const crs = layer.bbox?.crs || "EPSG:3857";
+    const { extent } = getExtentForProjection(crs);
+    const [, , maxX, maxY] = turfBbox({type: 'FeatureCollection', features: layer.features || []});
```
Comment (MV88) [inline web/client/components/import/dragZone/enhancers/useFiles.js:24]: why you skip minX and minY ?

File: web/client/components/import/dragZone/enhancers/useFiles.js [hunk 5]
```diff
@@ -1,5 +1,29 @@
...
+// retrieves projectionDefs
+const getProjections = () => {
+    const projections = (getConfigProp('projectionDefs') || []).concat([{code: "EPSG:3857", extent: [-20026376.39, -20048966.10, 20026376.39, 20048966.10]},
+        {code: "EPSG:4326", extent: [-180, -90, 180, 90]}
+    ]);
+    return projections;
+};
+
+// finds a projection from projectionDefs
+const getExtentForProjection = code => {
```
Comment (MV88) [inline web/client/components/import/dragZone/enhancers/useFiles.js:16]: i would use a default value for code like "EPSG:3857"

**Review by MV88 — CHANGES_REQUESTED** [review]
the pr solves the issue, but please refactor it a bit and add unit tests

please checkout my comments

File: web/client/localConfig.json (hunk 1 above)
Comment (MuhweziDeo) [reply web/client/localConfig.json:422]: Sorry was testing using this will revert this back

File: web/client/components/import/dragZone/enhancers/useFiles.js (hunk 4 above)
Comment (MuhweziDeo) [reply web/client/components/import/dragZone/enhancers/useFiles.js]: I used the maxX and minY to compare if the max X and Y for the layer can fit within the extent for the projection. Am I doing it wrong @MV88

File: build/tests.webpack.js [hunk 6]
```diff
@@ -1,3 +1,3 @@
-var context = require.context('../web', true, /(-test\.jsx?)|(-test-chrome\.jsx?)$/);
+var context = require.context('../web/client/utils', true, /(-test\.jsx?)|(-test-chrome\.jsx?)$/);
```
Comment (MV88) [inline build/tests.webpack.js:1]: don't push this

File: web/client/utils/CoordinatesUtils.js [hunk 7]
```diff
@@ -1030,6 +1033,37 @@ export const getPolygonFromCircle = (center, radius, units = "degrees", steps =
...
+
+/**
+ * Return a boolean to show if a layer fits within a boundary/extent
+ * @param layer {object} to check if fits with in a projection boundary
+ * @return {boolean} true or false
+ */
+export const checkIfLayerFitsExtentForProjection = (layer = {}) => {
+    const crs = layer.bbox?.crs || "EPSG:3857";
+    const { extent } = getExtentForProjection(crs);
+    const [, , maxX, maxY] = turfBbox({type: 'FeatureCollection', features: layer.features || []});
```
Comment (MV88) [inline web/client/utils/CoordinatesUtils.js:1064]: a layer boundary can exceed crs one also on min values, that's why the the whole extent should be contained in the crs one

**Review by MV88 — CHANGES_REQUESTED** [review]

File: web/client/utils/CoordinatesUtils.js [hunk 8]
```diff
@@ -1030,6 +1033,37 @@ export const getPolygonFromCircle = (center, radius, units = "degrees", steps =
...
+};
+
+/**
+ * Return a boolean to show if a layer fits within a boundary/extent
+ * @param layer {object} to check if fits with in a projection boundary
+ * @return {boolean} true or false
+ */
+export const checkIfLayerFitsExtentForProjection = (layer = {}) => {
+    const crs = layer.bbox?.crs || "EPSG:3857";
+    const { extent } = getExtentForProjection(crs);
```
Comment (MV88) [inline web/client/utils/CoordinatesUtils.js:1063]: probably is clearer
```suggestion
    const [crsMinX, crsMinY, crsMaxX,crsMaxY] = getExtentForProjection(crs).extent;
```

File: web/client/utils/CoordinatesUtils.js [hunk 9]
```diff
@@ -1030,6 +1033,37 @@ export const getPolygonFromCircle = (center, radius, units = "degrees", steps =
...
+    const [minx, minY, maxX, maxY] = turfBbox({type: 'FeatureCollection', features: layer.features || []});
+    return ((extent[2] >= maxX) && (extent[3] >= maxY) && (extent[1] >= minx) && (extent[2] > minY));
```
Comment (MV88) [inline web/client/utils/CoordinatesUtils.js:1065]: there is something wrong here
you are not using extent[0] and when comparing to max they should be inside
please add unit test since they are not validating the correct behaviour
add a unit test for a valid case  and many for invalid with only one condition that is failing
```suggestion
    return ((minx >= crsMinX) && (minY >= crsMinY) && (maxX <= crsMaxX) && (maxY <= crsMaxY));
```
please double check it

**Review by MV88 — CHANGES_REQUESTED** [review]
please review the code and the tests

File: web/client/utils/__tests__/CoordinatesUtils-test.js [hunk 10]
```diff
@@ -847,4 +848,42 @@ describe('CoordinatesUtils', () => {
...
+                    "type": "Point",
+                    "coordinates": [-150, 94]
+                },
+                "properties": {
+                    "prop0": "value0"
+                }
+            }]
+        };
+        const canFitWithBounds = checkIfLayerFitsExtentForProjection({name: "test", ...geoJson});
+        expect(canFitWithBounds).toBe(false);
```
Comment (MV88) [inline web/client/utils/__tests__/CoordinatesUtils-test.js:868]: how can this be outside bounds of 4326?

also create some other tests where part of the geometry is outside on each side of the bbox

**Review by MV88 — CHANGES_REQUESTED** [review]

File: web/client/utils/__tests__/CoordinatesUtils-test.js (hunk 10 above)
Comment (MuhweziDeo) [reply web/client/utils/__tests__/CoordinatesUtils-test.js:868]: for 4326  this is max extent`[-180, -90, 180, 90]`. so it's outside because the 94 exceeds the maximum of 90

**Review by MV88 — APPROVED** [review]

**Comment by MV88** [comment]
@ElenaGallo please test
